listings/applications/erc721/src/erc721.cairo

```diff
@@ -0,0 +1,193 @@
+        }
+
+        fn approve(ref self: ContractState, approved: ContractAddress, token_id: u256) {
+            let owner = self._require_owned(token_id);
+            let caller = get_caller_address();
+            assert(
+                caller == owner || self.is_approved_for_all(owner, caller), Errors::UNAUTHORIZED
+            );
+
+            self.approvals.write(token_id, approved);
+            self.emit(Approval { owner, approved, token_id });
+        }
+
+        fn get_approved(self: @ContractState, token_id: u256) -> ContractAddress {
+            self._require_owned(token_id);
+            self.approvals.read(token_id)
+        }
+
+        fn transfer_from(
+            ref self: ContractState, from: ContractAddress, to: ContractAddress, token_id: u256
+        ) {
+            let previous_owner = self._require_owned(token_id);
+            assert(from == previous_owner, Errors::INVALID_SENDER);
+            assert(!to.is_zero(), Errors::INVALID_RECEIVER);
+            assert(
+                self._is_approved_or_owner(from, get_caller_address(), token_id),
+                Errors::UNAUTHORIZED
+            );
+
+            self.balances.write(from, self.balances.read(from) - 1);
+            self.balances.write(to, self.balances.read(to) + 1);
+            self.owners.write(token_id, to);
+            self.approvals.write(token_id, Zero::zero());
+
+            self.emit(Transfer { from, to, token_id });
+        }
+
+        fn safe_transfer_from(
+            ref self: ContractState,
+            from: ContractAddress,
+            to: ContractAddress,
+            token_id: u256,
+            data: Span<felt252>
+        ) {
+            Self::transfer_from(ref self, from, to, token_id);
+            assert(
+                self._check_on_erc721_received(from, to, token_id, data),
+                Errors::SAFE_TRANSFER_FAILED
+            );
+        }
+
+        fn is_approved_for_all(
+            self: @ContractState, owner: ContractAddress, operator: ContractAddress
+        ) -> bool {
+            self.operator_approvals.read((owner, operator))
+        }
+
+        fn mint(ref self: ContractState, to: ContractAddress, token_id: u256) {
+            assert(!to.is_zero(), Errors::INVALID_RECEIVER);
+            assert(self.owners.read(token_id).is_zero(), Errors::ALREADY_MINTED);
+
+            self.balances.write(to, self.balances.read(to) + 1);
+            self.owners.write(token_id, to);
+
+            self.emit(Transfer { from: Zero::zero(), to, token_id });
+        }
+
+        fn burn(ref self: ContractState, token_id: u256) {
+            let owner = self._require_owned(token_id);
+
+            self.balances.write(owner, self.balances.read(owner) - 1);
+
+            self.owners.write(token_id, Zero::zero());
+            self.approvals.write(token_id, Zero::zero());
+
+            self.emit(Transfer { from: owner, to: Zero::zero(), token_id });
+        }
```

Comment:
Move these in the internal impl as `_mint` and `_burn`, and implement `mint` and `burn` in a separate interface impl block that call the internal functions